modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -114,32 +115,54 @@ object EnrichmentManager {
            )
              .leftMap(NonEmptyList.one)
              .possiblyExitingEarly(emitIncomplete)
-    } yield enriched
+    } yield (enriched, validContexts ::: extractResult.validationInfoContexts)
 
-    iorT.leftMap(_.head)
+    // derived contexts are set lastly because we want to include failure entities
+    // to derived contexts as well and we can get failure entities only in the end
+    // of the enrichment process
+    setDerivedContexts(iorT).leftMap(_.head.badRow).map(_._1)
   }
 
+  private def setDerivedContexts[F[_]: Sync](enriched: EnrichmentResult[F]): EnrichmentResult[F] =
+    IorT(
+      enriched.value.flatTap(v => Sync[F].delay {
+        val (derivedContexts, enriched) = v match {
+          case Ior.Right((e, l)) => (l, e.some)
+          case Ior.Left(l) => (extractFailureEntities(l), None)
+          case Ior.Both(b, (e, l)) => (l ::: extractFailureEntities(b), e.some)
+        }
+        for {
+          c <- ME.formatContexts(derivedContexts)
+          e <- enriched
+          _ = e.derived_contexts = c
+        } yield ()
+      })
+    )
```

Comment:
This is quite complicated

Reply:
I agree but that was the best I could come up with. Let me know if you have any idea to make it better.